source/game_sa/CPad.cpp

```diff
@@ -47,58 +64,139 @@ void CPad::ReconcileTwoControllersInput(CControllerState const& controllerA, CCo
 
 // Converted from thiscall void CPad::SetDrunkInputDelay(int delay) 0x53F910
 void CPad::SetDrunkInputDelay(int delay) {
-    plugin::CallMethod<0x53F910, CPad*, int>(this, delay);
+    DrunkDrivingBufferUsed = delay;
 }
 
-// Converted from thiscall void CPad::StartShake(short time,uchar frequency,uint) 0x53F920
-void CPad::StartShake(short time, unsigned char frequency, unsigned int arg2) {
-    plugin::CallMethod<0x53F920, CPad*, short, unsigned char, unsigned int>(this, time, frequency, arg2);
+// Converted from thiscall void CPad::StartShake(short time,uchar freq,uint) 0x53F920
+void CPad::StartShake(short time, unsigned char freq, unsigned int shakeDelayMs) {
+    if (!FrontEndMenuManager.field_20 || CCutsceneMgr::ms_running) {
+        return;
+    }
+    if (freq) {
+        if (CTimer::m_snTimeInMilliseconds >= NoShakeBeforeThis || freq > NoShakeFreq) {
+            if (ShakeDur) {
+                ShakeDur = time;
+                ShakeFreq = freq;
+            }
+            NoShakeBeforeThis = shakeDelayMs + CTimer::m_snTimeInMilliseconds;
+            NoShakeFreq = freq;
+        } 
+    } else {
+        ShakeDur = 0;
+        ShakeFreq = 0;
+    }
 }
 
-// Converted from thiscall void CPad::StartShake_Distance(short time,uchar frequency,float x,float y,float z) 0x53F9A0
-void CPad::StartShake_Distance(short time, unsigned char frequency, float x, float y, float z) {
-    plugin::CallMethod<0x53F9A0, CPad*, short, unsigned char, float, float, float>(this, time, frequency, x, y, z);
+// 0x53F9A0
+void CPad::StartShake_Distance(short time, unsigned char freq, CVector pos) {
+    if (!FrontEndMenuManager.field_20 || CCutsceneMgr::ms_running) {
+        return;
+    }
+    if ((TheCamera.GetPosition() - pos).Magnitude() < 70.0f) {
+        if (freq) {
+            if (time > ShakeDur) {
+                ShakeDur = time;
+                ShakeFreq = freq;
+            }
+        } else {
+            ShakeDur = 0;
+            ShakeFreq = 0;
+        }
+    }
+
 }
 
-// Converted from thiscall void CPad::StartShake_Train(float x,float y) 0x53FA70
-void CPad::StartShake_Train(float x, float y) {
-    plugin::CallMethod<0x53FA70, CPad*, float, float>(this, x, y);
+// 0x53FA70
+void CPad::StartShake_Train(CVector2D point) {
+    if (!FrontEndMenuManager.field_20 || CCutsceneMgr::ms_running) {
+        return;
+    }
+    if (FindPlayerVehicle(-1) && FindPlayerVehicle(-1)->m_vehicleType == 6) {
+        return;
+    }
+    const auto fDistSq = DistanceBetweenPointsSquared(TheCamera.GetPosition(), point);
+    if (ShakeDur < 100 && fDistSq < 4900.0f) {
+        ShakeDur = 100;
+        ShakeFreq = static_cast<char>(30.0f + 70.0f - sqrt(fDistSq));
```

Comment:
Yeah, the order does not matter, but let it be as in the original.

```suggestion
        ShakeFreq = static_cast<char>(70.0f - sqrt(fDistSq) + 30.0f);
```